Initial Implementation for Runtime Backpressure

Theodus opened the pull request:
This is an initial version of the runtime backpressure system.

The primary goal of backpressure is to prevent runaway growth of a cown's message queue that may eventually result in a Verona application running out of memory. This system is loosely based on Pony's generalized runtime backpressure. Two tests are included that, without backpressure, would either run out of memory or complete with a much longer total runtime.

Closes #33

rengolin commented:
Is there a paper / document / post about how this approach solves backpressure? Reviewing the code without a good understanding of the process itself is hard. Even if it's Pony specific, some context would be good.

Theodus commented:
@rengolin Here are some resources for Pony backpressure for context:
- [Pony stdlib backpressure package docs](https://stdlib.ponylang.io/backpressure--index/)
- [Initial Pony RFC for what would become the full backpressure system](https://github.com/ponylang/rfcs/blob/master/text/0017-tcp-backpressure.md)

I'll follow up soon with an overview of how this differs from the Pony backpressure system.

SeanTAllen commented:
Neither of those really cover the runtime aspects of the Pony backpressure system however. That lives in comments in the Pony runtime codebase (as I have never gotten around to writing it up).

Here's a copy and paste of my answering the basics question for someone a couple months back:

There's no credit system. The overhead is too high performance wise. Too much message passing and I couldn't find a way to make it work for "extreme fan in" situations where you have hundreds of actors sending to a single actor.

The high-level idea is this:

1. you use a heuristic to decide if an actor is overloaded (more details on specifics for pony right now)
2. if an actor is overloaded and a message is sent to it. then the sending actor is "muted" unless the sending actor is itself overloaded.
3. muted actors are descheduled "as soon as possible" and will not be rescheduled until such time that the actor that caused it to be muted is no longer overloaded.
This works well, but there are negative performance implications in "thundering herd scenarios" where things get muted that really don't need to be. We are going to be testing soon a change that instead of "immediately muting", there's a bit of a ramp up. You can see a PR for that here:

https://github.com/ponylang/ponyc/pull/3009

The current "actor is overloaded" heuristic is based on Pony having a cooperative scheduler that is based on number of messages rather than time slicing to give an actor time on a scheduler. By default, each Pony actor can process up to 100 messages before it is descheduled. The heuristic in place is "an actor is overloaded if there are messages remaining on its queue when its done with a run". Which translates to "had at least 101 message in its queue".

There's very little message passing with this system and we found, unlike a credit based system, that there is no performance impact (that is noticeable in real world, millions of messages a second application).

That's the basics. Fire away with any questions you might have.

Theodus commented:
The main difference here is that the heuristic used is somewhat similar to that of fairness, where a token occasionally falls out of the queue. This given an approximation for how many messages have been processed since the token last fell out. Details on how this works in practice are described in [backpressure.h](https://github.com/microsoft/verona/pull/172/files#diff-a5ad9506408b758ab699d027d3ce25e0R7) and [this part of Cown::run](https://github.com/microsoft/verona/pull/172/files?file-filters%5B%5D=.h#diff-97a2486aafb2256800bdee37754f2c94R931-R954).

Theodus commented:
Cowns are then muted once an 8-bit counter called `pressure` maxes out. The pressure is propagated from receivers to senders in [Cown::backpressure_scan](https://github.com/microsoft/verona/blob/1fda72c959438fdf02d7bf1d0286524c6180fc01/src/rt/sched/cown.h#L788-L808) on every message send.

Theodus commented:
Additionally, here is a collection of comments from the Pony runtime that I accumulated a few months ago to help wrap my head around it:


___
[actor.c:228](https://github.com/ponylang/ponyc/blob/8d2d492858deea7048e188e503f99697931bb206/src/libponyrt/actor/actor.c#L228)

if we become muted as a result of handling a message, bail out now. we aren't set to "muted" at this point. setting to muted during a behavior can lead to race conditions that might result in a deadlock. Given that actor's are not run when they are muted, then when we started our batch, actor->muted would have been 0. If any of our message sends would result in the actor being muted, that will have changed to greater than 0.

We will then set the actor to "muted". Once set, any actor sending a message to it will be also be muted unless said sender is marked as overloaded.

The key points here is that:
1. We can't set the actor to "muted" until after its running a behavior.
2. We should bail out from running the actor and return false that it won't be rescheduled.
___
[actor.c:256](https://github.com/ponylang/ponyc/blob/8d2d492858deea7048e188e503f99697931bb206/src/libponyrt/actor/actor.c#L256)

If we hit our batch size, consider this actor to be overloaded. Overloaded actors are allowed to send to other overloaded actors and to muted actors without being muted themselves.
___
[actor.c:279](https://github.com/ponylang/ponyc/blob/8d2d492858deea7048e188e503f99697931bb206/src/libponyrt/actor/actor.c#L279)

if we were overloaded and didn't process a full batch, set ourselves as no longer overloaded. Once this is done:
1. sending to this actor is no longer grounds for an actor being muted
2. this actor can no longer send to other actors free from muting should the receiver be overloaded or muted
___
[actor.c:545](https://github.com/ponylang/ponyc/blob/8d2d492858deea7048e188e503f99697931bb206/src/libponyrt/actor/actor.c#L545)

only mute a sender if all of the following are true:
1. the receiver is overloaded/under pressure/muted
2. the sender isn't overloaded
3. we are sending to another actor (as compared to sending to self)
___
[actor.c:743](https://github.com/ponylang/ponyc/blob/8d2d492858deea7048e188e503f99697931bb206/src/libponyrt/actor/actor.c#L743)

Mute/Unmute/Check mute status functions

For backpressure related muting and unmuting to work correctly, the following rules have to be maintained.
1. Across schedulers, an actor should never been seen as muted when it is not in fact muted.
2. It's ok for a muted actor to be seen as unmuted in a transient fashion across actors

If rule #1 is violated, we might end up deadlocking because an actor was muted for sending to an actor that might never be unmuted (because it isn't muted). The actor muted actor would continue to remain muted and the actor incorrectly seen as muted became actually muted and then unmuted.

If rule #2 is violated, then a muted actor will receive from 1 to a few additional messages and the sender won't be muted. As this is a transient situtation that should be shortly rectified, there's no harm done. Our handling of atomic operations in `ponyint_is_muted` and `ponyint_unmute_actor` are to assure that rule #1 isn't violated. We have far more relaxed usage of atomics in `ponyint_mute_actor` given the far more relaxed rule #2.

An actor's `is_muted` field is effectly a `bool` value. However, by using a `uint8_t`, we use the same amount of space that we would for a boolean but can use more efficient atomic operations. Given how often these methods are called (at least once per message send), efficiency is of primary importance.
___
[scheduler.c:237](https://github.com/ponylang/ponyc/blob/8d2d492858deea7048e188e503f99697931bb206/src/libponyrt/sched/scheduler.c#L237)

Only send block message if we don't have any muted actors. If we have at least one muted actor it means we aren't really blocked. There's work that can eventually be done.
...
and actor was unmuted and added to our run queue. pop it and return. effectively, we are "stealing" from ourselves
...
Someone else stole from our newly unmuted actor. If we have no more muted actors, we need to inform everyone that we are blocked
...
In response to reading a message, we might have unmuted an actor and added it back to our queue. if we don't have an actor to run, we want to pop from our queue to check for a recently unmuted actor
___
[scheduler.c:509](https://github.com/ponylang/ponyc/blob/8d2d492858deea7048e188e503f99697931bb206/src/libponyrt/sched/scheduler.c#L509)

Manage a scheduler's mute map

When an actor attempts to send to an overloaded actor, it will be added to the mute map for this scheduler. The mute map is in the form of:
```
overloaded receiving actor => [sending actors]
```
- A given actor will only existing as a sending actor in the map for a given scheduler.
- Receiving actors can exist as a mute map key in the mute map of more than one scheduler

Because muted sending actors only exist in a single scheduler's mute map and because they aren't scheduled when they are muted, any manipulation that we do on their state (for example incrementing or decrementing their mute count) is thread safe as only a single scheduler thread will be accessing the information.

rengolin commented:
Sean/Theo, thanks for all the references, it helped a lot.

IIUC, the system is basically a priority queue which temporarily blocks producers in cases that would increase backpressure. I'm assuming this works across different queues (ie threads running on separate cores can increase the overall system's volume but still being treated as silos of pressure). Across whole systems (different machines) the backpressure would piggy back on TCP's (or IB, RDMA, whatever) own backpressure systems and close the gap.

The number of messages seems like an odd choice for the measurement of pressure, however. A small system could cope with less than others, or larger messages could be harder to handle. Waiting time has been traditionally used for that kind of metric, but that also has it's own problems.

In the future, it would be interesting to allow the actor to define its own pressure and just query that directly, but I have no specific case in mind to back this up, so I'll just leave it here as a topic to think about. May be a completely bogus idea.

rengolin commented on src/rt/sched/backpressure.h:
```diff
@@ -0,0 +1,130 @@
+   * Calculate the total load accumulated for this cown. This value will be in
+   * the range [0, 1215].
+   */
+  static inline uint32_t load(uint32_t bp)
```
nit: `load` to me indicates an action "to load a value". Perhaps `get_load` would be clearer to other compiler engineers as well.

rengolin commented on src/rt/sched/backpressure.h:
```diff
@@ -0,0 +1,130 @@
+  }
+
+  /**
+   * Return if enough pressure has been applied to mute the cown and the cown
```
Return `true` if...

Theodus commented:
> The number of messages seems like an odd choice for the measurement of pressure, however. A small system could cope with less than others, or larger messages could be harder to handle. Waiting time has been traditionally used for that kind of metric, but that also has it's own problems.

I assume you mean a measurement of "load". An important point is that the measurement is not of messages processed in a given time period. It is an approximation of how deep the messages queue has been. Cowns processing few messages for long periods of time will not trigger any pressure unless they are receiving more messages than they could process by the time the next token falls out.

rengolin commented on src/rt/sched/backpressure.h:
```diff
@@ -0,0 +1,130 @@
+  }
+
+  /**
+   * Distribute load from first participant of a multimessage. This is done
```
This comments the intended use, not the actual behaviour of the function. It'd be clearer to document this on the call site and have a generic comment here saying what ORring the two loads accomplish.

Theodus commented on src/rt/sched/schedulerthread.h:
```diff
@@ -168,6 +182,71 @@ namespace verona::rt
+      if (overloaded.size() != 0)
+        avg_load /= overloaded.size();
+
+      if (avg_load < Backpressure::unmute_threshold)
```
That would be the case if the average is pulled down below the threshold by a large amount of cowns that quickly fall from the overload threshold down to near zero. An alternative may be to unmute when no cowns in the set are above the threshold. An upside to that option would be the ability to short circuit this scan as soon as a cown with a load above the threshold is found.

Theodus commented on src/rt/sched/cown.h:
```diff
@@ -771,11 +775,67 @@ namespace verona::rt
+    bool apply_backpressure(Alloc* alloc, MessageBody* senders)
+    {
+      bool muted = false;
+      const auto bp0 =
+        senders->cowns[0]->backpressure.load(std::memory_order_relaxed);
```
The load on the first cown is tracked normally, however the other participants often have their load reset whenever they are acquired for the message. This means that, for example, the receivers in `perf-cowngc2` all give off a load signal that is too weak to trigger adequate backpressure since the receivers are constantly participating in different multimessages and shuffling around which one is at index 0.

Theodus replied:
So the load is distributed from the first cown.

Theodus commented on src/rt/sched/backpressure.h:
```diff
@@ -0,0 +1,139 @@
+     * |31          24|23              8|7      0|
+     * |   pressure   |   load history  |  load  |
+     *
+     * The 8-bit pressure field is 0 when muted and in the range [1, 255]
+     * otherwise.
```
Pressure is a sum of messages sent to other cowns where each count is wighted by the load on the receiving cown.

Theodus replied:
More explanation has been added to the comment

Theodus commented on src/rt/sched/backpressure.h:
```diff
@@ -0,0 +1,139 @@
+    inline void distribute_load(const Backpressure& bp0)
+    {
+      bits |= bp0.current_load();
+    }
```
Comment added

Theodus commented on src/rt/sched/schedulerthread.h:
```diff
@@ -83,13 +85,22 @@ namespace verona::rt
     size_t total_cowns = 0;
     std::atomic<size_t> free_cowns = 0;
 
+    typename T::MessageBody* message_body = nullptr;
```
Comment added.

Theodus commented on src/rt/sched/schedulerthread.h:
```diff
@@ -83,13 +85,22 @@ namespace verona::rt
 
+    typename T::MessageBody* message_body = nullptr;
+
+    ObjectMap<T*> muted;
```
You are correct. Comments added.

Theodus commented on src/rt/sched/schedulerthread.h:
```diff
@@ -83,13 +85,22 @@ namespace verona::rt
+    typename T::MessageBody* message_body = nullptr;
+
+    ObjectMap<T*> muted;
+    ObjectMap<T*> overloaded;
```
You are correct. Comments added.

Theodus commented on src/rt/sched/cown.h:
```diff
@@ -771,11 +775,67 @@ namespace verona::rt
+        auto* receiver = receivers.cowns[r];
+        const auto bp = receiver->backpressure.load(std::memory_order_acquire);
+        const auto load = bp.total_load();
+        pressure += load >> 3;
+        if (load > Backpressure::overload_threshold)
+          Scheduler::local()->overload(receiver);
```
Comment added.

Theodus commented:
Thanks for the feedback! I'm closing for now since I think there is a better approach to unmuting without risking deadlock.
